packages/react-components/react-color-picker-preview/library/src/components/ColorArea/useColorArea.ts

```diff
@@ -9,23 +18,143 @@ import type { ColorAreaProps, ColorAreaState } from './ColorArea.types';
+      return;
+    }
+    const coordinates = getCoordinates(rootRef.current, event);
+    const newColor: HsvColor = {
+      ...color,
+      s: coordinates.x,
+      v: coordinates.y,
+    };
+
+    setColor(newColor);
+    onChange?.(event, { type: 'change', event, color: newColor });
+  });
+
+  const handleDocumentMouseMove = React.useCallback(
+    (event: MouseEvent) => {
+      requestColorChange(event);
+    },
+    [requestColorChange],
+  );
+  const handleDocumentMouseUp = useEventCallback(() => {
+    targetDocument?.removeEventListener('mousemove', handleDocumentMouseMove);
+  });
+
+  const _onMouseDown: React.MouseEventHandler<HTMLDivElement> = useEventCallback(event => {
+    event.stopPropagation();
+    event.preventDefault();
+
+    requestColorChange(event.nativeEvent);
+
+    targetDocument?.addEventListener('mousemove', handleDocumentMouseMove);
+    targetDocument?.addEventListener('mouseup', handleDocumentMouseUp, { once: true });
+  });
+
+  const _onChange: React.ChangeEventHandler<HTMLInputElement> = useEventCallback(event => {
+    const newColor = { h: color.h, s: Number(xRef.current!.value) / 100, v: Number(yRef.current?.value) / 100, a: 1 };
+
+    setColor(newColor);
+    onChange?.(event, {
+      type: 'change',
+      event,
+      color: newColor,
+    });
+  });
+
+  const _onKeyDown = useEventCallback((event: React.KeyboardEvent<HTMLInputElement>) => {
+    if (event.key === 'ArrowUp' || event.key === 'ArrowDown') {
+      event.preventDefault();
+
+      const newY = event.key === 'ArrowUp' ? clamp(value + 1, MIN, MAX) : clamp(value - 1, MIN, MAX);
+      const newColor: HsvColor = {
+        ...color,
+        s: color.s,
+        v: newY / 100,
+      };
+      setColor(newColor);
+      onChange?.(event, { type: 'change', event, color: newColor });
+    }
+  });
+
+  const rootVariables = {
+    [colorAreaCSSVars.areaXProgressVar]: `${saturation}%`,
+    [colorAreaCSSVars.areaYProgressVar]: `${value}%`,
+    [colorAreaCSSVars.thumbColorVar]: 'transparent',
+    [colorAreaCSSVars.mainColorVar]: `hsl(${color.h}, 100%, 50%)`,
+  };
+
+  const state: ColorAreaState = {
     components: {
-      // TODO add each slot's element type or component
+      inputX: 'input',
+      inputY: 'input',
       root: 'div',
+      thumb: 'div',
     },
-    // TODO add appropriate slots, for example:
-    // mySlot: resolveShorthand(props.mySlot),
-    root: slot.always(
-      getIntrinsicElementProps('div', {
-        ref,
-        ...props,
-      }),
-      { elementType: 'div' },
-    ),
+    root: slot.always(getIntrinsicElementProps('div', { ...rest, ref }), { elementType: 'div' }),
+    inputX: slot.always(inputX, {
+      defaultProps: {
+        id: useId('sliderX-', props.id),
+        type: 'range',
+        ref: xRef,
+      },
+      elementType: 'input',
+    }),
+    inputY: slot.always(inputY, {
+      defaultProps: {
+        id: useId('sliderY-', props.id),
+        type: 'range',
+        ref: yRef,
```

Comment:
If you need to pass slots to refs, you need to do `state.SLOT.ref = useMergedRefs(state.SLOT.ref, REF);` so users' input won't override it

Reply:
fixed